File: src/OSMdata.R
###############
# Introduction: data recovering
setwd("/home/rde/Documents/osland-ia")
tlnodes = read.csv("/home/rde/data/latest-bordeaux-metropole-node-timeline.csv",stringsAsFactors = F)
tlways = read.csv("/home/rde/data/latest-bordeaux-metropole-way-timeline.csv",stringsAsFactors = F)
tlrelations = read.csv("/home/rde/data/latest-bordeaux-metropole-relation-timeline.csv",stringsAsFactors = F)
save.image("/home/rde/Documents/osland-ia/data/bordeaux-metropole.RData")
#
View(tlnodes)
View(tlways)
View(tlrelations)
#
tlnodes$ts = as.POSIXct(tlnodes$ts)
tlways$ts = as.POSIXct(tlways$ts)
tlrelations$ts = as.POSIXct(tlrelations$ts)
###############
### Recover last version of each element
updatedelem = function(data){
  # Recover last version of nodes
  upddata = aggregate(version~id , data=data , FUN=max)
  upddata = merge(data,upddata)
  upddata = upddata[order(upddata$id),]
}
nodes = updatedelem(tlnodes)
ways = updatedelem(tlways)
relations = updatedelem(tlrelations)
###############

###############
### Introduction: building of the global metadata: data history with respect to user and change set ids
useractivity = cbind(elem=rep("n",nrow(tlnodes)),tlnodes[,c("id","version","uid","chgset","ts","ntags")])
useractivity = rbind( useractivity , cbind(elem=rep("w",nrow(tlways)),tlways[,c("id","version","uid","chgset","ts","ntags")]) )
useractivity = rbind( useractivity , cbind(elem=rep("r",nrow(tlrelations)),tlrelations[,c("id","version","uid","chgset","ts","ntags")]) )
useractivity = useractivity[order(useractivity$ts),]
str(useractivity)
# proportion of each element in the data history
table(useractivity$elem)

### Analyse 1: how many modifications per change set
modifbychgset = aggregate(cbind(elem,id)~chgset , data=useractivity, FUN=length)
modifbychgset = merge(modifbychgset[,-3] , unique(useractivity[,c("chgset","uid")]))[,c(1,3,2)]
head(modifbychgset)
plot(quantile(modifbychgset$elem,probs=seq(0,1,0.001)),seq(0,1,0.001),type="l",col="red",lwd=2,xlim=c(0,100),xlab="Modification quantity per change set",ylab="Empirical CDF")
abline(h=c(0,1),v=0)
nodemodifbychgset = setNames(aggregate(id~chgset , data=useractivity[useractivity$elem=="n",], FUN=length),c("chgset","nnodemodif"))
waymodifbychgset = setNames(aggregate(id~chgset , data=useractivity[useractivity$elem=="w",], FUN=length),c("chgset","nwaymodif"))
relationmodifbychgset = setNames(aggregate(id~chgset , data=useractivity[useractivity$elem=="r",], FUN=length),c("chgset","nrelationmodif"))
plot(quantile(nodemodifbychgset$nnodemodif,probs=seq(0,1,0.001)),seq(0,1,0.001),type="l",col="red",lwd=2,xlim=c(0,100),xlab="Modification per change set",ylab="Empirical CDF")
lines(quantile(waymodifbychgset$nwaymodif,probs=seq(0,1,0.001)),seq(0,1,0.001),col="blue",lwd=2)
lines(quantile(relationmodifbychgset$nrelationmodif,probs=seq(0,1,0.001)),seq(0,1,0.001),col="green",lwd=2)
abline(h=c(0,1),v=0)
legend("bottomright",legend=c("nodes","ways","relations"),col=c("red","blue","green"),lwd=2)
chgsetsynthesis = merge(modifbychgset,nodemodifbychgset,all=T)
chgsetsynthesis = merge(chgsetsynthesis,waymodifbychgset,all=T)
chgsetsynthesis = merge(chgsetsynthesis,relationmodifbychgset,all=T)
remove(nodemodifbychgset,waymodifbychgset,relationmodifbychgset)

# Analyse 2: how many change set per user
chgsetbyuser = aggregate(chgset~uid,data=modifbychgset,FUN=length)
head(chgsetbyuser)
quantile(chgsetbyuser$chgset,probs=seq(0,1,0.1))
plot(quantile(chgsetbyuser$chgset,probs=seq(0,1,0.001)),seq(0,1,0.001),type="l",col="red",xlim=c(0,100),xlab="Change set quantity per user",ylab="Empirical CDF")
abline(h=c(0,1),v=0)

# Analyse 3: how many modification per user
modifbyuser = aggregate(cbind(elem,id)~uid , data=useractivity, FUN=length)[,-3]
plot(quantile(modifbychgset$elem,probs=seq(0,1,0.001)),seq(0,1,0.001),type="l",col="red",lwd=2,xlim=c(0,100),xlab="Total modification per user",ylab="Empirical CDF")
abline(h=c(0,1),v=0)
head(modifbyuser)
usersynthesis = merge(chgsetbyuser,modifbyuser)
remove(chgsetbyuser,modifbyuser)
plot(usersynthesis$chgset,usersynthesis$elem,col="blue",pch=15,cex=0.5,xlab="Number of change set",ylab="Number of modified elements",main="User mapping")
abline(v=0,h=0)
plot(usersynthesis$chgset,usersynthesis$elem,col="blue",pch=15,cex=0.5,xlab="Number of change set",ylab="Number of modified elements",main="User mapping",xlim=c(0,1000),ylim=c(0,10000))
abline(v=0,h=0)
nodemodifperuser = setNames(aggregate(id~uid , data=useractivity[useractivity$elem=="n",], FUN=length),c("uid","nnodemodif"))
waymodifperuser = setNames(aggregate(id~uid , data=useractivity[useractivity$elem=="w",], FUN=length),c("uid","nwaymodif"))
relationmodifperuser = setNames(aggregate(id~uid , data=useractivity[useractivity$elem=="r",], FUN=length),c("uid","nrelationmodif"))
plot(quantile(nodemodifperuser$nnodemodif,probs=seq(0,1,0.001)),seq(0,1,0.001),type="l",col="red",lwd=2,xlim=c(0,200),xlab="Modification per user",ylab="Empirical CDF")
lines(quantile(waymodifperuser$nwaymodif,probs=seq(0,1,0.001)),seq(0,1,0.001),col="blue",lwd=2)
lines(quantile(relationmodifperuser$nrelationmodif,probs=seq(0,1,0.001)),seq(0,1,0.001),col="green",lwd=2)
abline(h=c(0,1),v=0)
legend("bottomright",legend=c("nodes","ways","relations"),col=c("red","blue","green"),lwd=2)
usersynthesis = merge(usersynthesis,nodemodifperuser,all=T)
usersynthesis = merge(usersynthesis,waymodifperuser,all=T)
usersynthesis = merge(usersynthesis,relationmodifperuser,all=T)

# Analyse 4: how many user per elements
userbyelem = aggregate(uid~elem+id,data=useractivity,FUN=function(x){length(unique(x))})
head(userbyelem)
plot(quantile(userbyelem[userbyelem$elem=="n","uid"],probs=seq(0,1,0.001)),seq(0,1,0.001),type="l",col="red",lwd=2,xlim=c(0,20),xlab="Unique contributor quantity per element",ylab="Empirical CDF")
lines(quantile(userbyelem[userbyelem$elem=="w","uid"],probs=seq(0,1,0.001)),seq(0,1,0.001),col="blue",lwd=2)
lines(quantile(userbyelem[userbyelem$elem=="r","uid"],probs=seq(0,1,0.001)),seq(0,1,0.001),col="green",lwd=2)
abline(h=c(0,1),v=0)
boxplot(userbyelem$uid~userbyelem$elem,ylim=c(0,10),col=c("pink","lightblue","lightgreen"),border=c("red","blue","green"))
par(mfrow=c(3,1))
hist(userbyelem[userbyelem$elem=="n","uid"],breaks=seq(0,130,1),xlim=c(0,10),ylim=c(0,1),freq=F,col="pink",border="red",main="",xlab="Number of unique contributors (nodes)")
hist(userbyelem[userbyelem$elem=="w","uid"],breaks=seq(0,130,1),xlim=c(0,10),ylim=c(0,1),freq=F,col="lightblue",border="blue",main="",xlab="Number of unique contributors (ways)")
hist(userbyelem[userbyelem$elem=="r","uid"],breaks=seq(0,130,1),xlim=c(0,10),ylim=c(0,1),freq=F,col="lightgreen",border="green",main="",xlab="Number of unique contributors (relations)")
par(mfrow=c(1,1))

# Analyse 5: how many version per elements
versionbyelem = aggregate(version~elem+id,data=useractivity,length)
head(versionbyelem)
elemsynthesis = merge(userbyelem,versionbyelem)
remove(userbyelem,versionbyelem)
plot(elemsynthesis$uid,elemsynthesis$version,col=elemsynthesis$elem,pch=3,cex=0.5,xlab="Number of unique contributors",ylab="Number of versions",main="Element building",ylim=c(0,600))
lines(1:2000,1:2000,lty="dotted",col="grey")
abline(v=0,h=0)
legend("topleft",legend=c("nodes","ways","relations"),col=1:3,pch=3)

# Analyse 6 : how many contributions per element and per user
contribbyuserelem = aggregate(version~elem+id+uid, data=useractivity, length)
head(contribbyuserelem)
plot(quantile(contribbyuserelem[contribbyuserelem$elem=="n","version"],probs=seq(0,1,0.001)),seq(0,1,0.001),type="l",col="red",lwd=2,xlim=c(0,10),xlab="Contribution per user per element",ylab="Empirical CDF")
lines(quantile(contribbyuserelem[contribbyuserelem$elem=="w","version"],probs=seq(0,1,0.001)),seq(0,1,0.001),col="blue",lwd=2)
lines(quantile(contribbyuserelem[contribbyuserelem$elem=="r","version"],probs=seq(0,1,0.001)),seq(0,1,0.001),col="green",lwd=2)
abline(h=c(0,1),v=0)
legend("bottomright",legend=c("nodes","ways","relations"),col=c("red","blue","green"),lwd=2)
nmodifperuserperelem = aggregate(version~uid, data=contribbyuserelem, median)
nnodemodifperuserperelem = setNames(aggregate(version~uid, data=contribbyuserelem[contribbyuserelem$elem=="n",], median),c("uid","nnodemodifperelem"))
nwaymodifperuserperelem = setNames(aggregate(version~uid, data=contribbyuserelem[contribbyuserelem$elem=="w",], median),c("uid","nwaymodifperelem"))
nrelationmodifperuserperelem = setNames(aggregate(version~uid, data=contribbyuserelem[contribbyuserelem$elem=="r",], median),c("uid","nrelationmodifperelem"))
usersynthesis = merge(usersynthesis,nmodifperuserperelem,all=T)
usersynthesis = merge(usersynthesis,nnodemodifperuserperelem,all=T)
usersynthesis = merge(usersynthesis,nwaymodifperuserperelem,all=T)
usersynthesis = merge(usersynthesis,nrelationmodifperuserperelem,all=T)
remove(nmodifperuserperelem,nnodemodifperuserperelem,nwaymodifperuserperelem,nrelationmodifperuserperelem)

# Analyse 7 : temporal description of user contributions
userfirst = setNames(aggregate(ts~uid , useractivity , min),c("uid","first"))
userlast = setNames(aggregate(ts~uid , useractivity , max),c("uid","last"))
usersynthesis = merge( usersynthesis , userfirst , all=T )
usersynthesis = merge( usersynthesis , userlast , all=T )
usersynthesis$activity.days = apply(usersynthesis , 1 , function(x){round(difftime(x[["last"]] , x[["first"]] , units="days")[[1]],2)})
plot(usersynthesis$nchgset , usersynthesis$activity.days , col="red",pch=3,cex=0.5,xlim=c(0,100),xlab="Change set per user",ylab="User activity (days)")
abline(v=0,h=0)

# Analyse 8 : temporal description of change sets
chgsetbegin = setNames(aggregate(ts~chgset , useractivity , min),c("chgset","begin"))
chgsetend = setNames(aggregate(ts~chgset , useractivity , max),c("chgset","end"))
chgsetsynthesis = merge( chgsetsynthesis , chgsetbegin , all=T )
chgsetsynthesis = merge( chgsetsynthesis , chgsetend , all=T )
chgsetsynthesis$duration = apply(chgsetsynthesis , 1 , function(x){round(difftime(x[["end"]] , x[["begin"]] , units="min")[[1]],2)})
plot(chgsetsynthesis$nnodemodif , chgsetsynthesis$duration , col="red",pch=3,cex=0.5,log="x",xlab="Modif per change set",ylab="Change set duration (min)")
points(chgsetsynthesis$nwaymodif , chgsetsynthesis$duration , col="blue",pch=3,cex=0.5)
points(chgsetsynthesis$nrelationmodif , chgsetsynthesis$duration , col="green",pch=3,cex=0.5,log="x")
abline(v=0,h=0)
legend("topright",legend=c("nodes","ways","relations"),col=c("red","blue","green"),pch=3)

# Analyse 9 : temporal description of element life cycle
elembegin = setNames(aggregate(ts~elem+id , useractivity , min),c("elem","id","creation"))
elemend = setNames(aggregate(ts~elem+id , useractivity , max),c("elem","id","lastmodif"))
elemsynthesis_ = elemsynthesis
elemsynthesis = merge( elemsynthesis , elembegin )
elemsynthesis = merge( elemsynthesis , elemend )
elemsynthesis$lifecycle = apply(elemsynthesis , 1 , function(x){round(difftime(x[["lastmodif"]] , x[["creation"]] , units="days")[[1]],2)})
elemsynthesis$available = rep(T,nrow(elemsynthesis))
delnodes = nodes[nodes$lon==Inf,"id"]
delways = ways[ways$nnodes==0,"id"]
delrelations = relations[relations$nmembers==0,"id"]
elemsynthesis[elemsynthesis$elem=="n"&elemsynthesis$id%in%delnodes,"available"] = F
elemsynthesis[elemsynthesis$elem=="w"&elemsynthesis$id%in%delways,"available"] = F
elemsynthesis[elemsynthesis$elem=="r"&elemsynthesis$id%in%delrelations,"available"] = F
timehorizon = as.POSIXct("2017-02-13 0:00:00")
elemsynt = elemsynthesis
elemsynthesis[elemsynthesis$available,"lifecycle"] = apply(elemsynthesis[elemsynthesis$available,] , 1 , function(x){round(difftime(timehorizon , x[["creation"]] , units="days")[[1]],2)})
elemsynthesis$lifeend = rep(timehorizon,nrow(elemsynthesis))
elemsynthesis[!elemsynthesis$available,"lifeend"] = elemsynthesis[!elemsynthesis$available,"lastmodif"]
remove(elemsynthesis_,elembegin,elemend)

### Summary: build a metadata structure for users, change sets and elements
# users
head(usersynthesis)
names(usersynthesis) = c("uid","nchgset","nmodif","nnodemodif","nwaymodif","nrelationmodif","nmodifperelem","nnodemodifperelem","nwaymodifperelem","nrelationmodifperelem","first","last","activity.days")
# change sets
head(chgsetsynthesis)
names(chgsetsynthesis) = c("chgset","uid","nmodif","nnodemodif","nwaymodif","nrelationmodif","begin","end","duration.min")
# elements
head(elemsynthesis)
names(elemsynthesis) = c("type","id","nuniqcontrib","nversion","creation","lastmodif","available","lifecycle.days","lifeend")

### Saving: keep important structures in .RData and .csv formats
save.image("/home/rde/Documents/osland-ia/data/bordeaux-metropole.RData")
#
write.csv(usersynthesis,"/home/rde/Documents/osland-ia/data/bordeaux-metropole-users.csv")
write.csv(chgsetsynthesis,"/home/rde/Documents/osland-ia/data/bordeaux-metropole-changesets.csv")
write.csv(elemsynthesis,"/home/rde/Documents/osland-ia/data/bordeaux-metropole-elements.csv")
write.csv(useractivity,"/home/rde/Documents/osland-ia/data/bordeaux-metropole-fullactivity.csv")
#
write.csv(nodes,"/home/rde/Documents/osland-ia/data/bordeaux-metropole-nodes.csv")
write.csv(ways,"/home/rde/Documents/osland-ia/data/bordeaux-metropole-ways.csv")
write.csv(relations,"/home/rde/Documents/osland-ia/data/bordeaux-metropole-relations.csv")
#
write.csv(tlnodes,"/home/rde/Documents/osland-ia/data/bordeaux-metropole-node-history.csv")
write.csv(tlways,"/home/rde/Documents/osland-ia/data/bordeaux-metropole-way-history.csv")
write.csv(tlrelations,"/home/rde/Documents/osland-ia/data/bordeaux-metropole-relation-history.csv")

File: src/OSMdata.py
# coding: utf-8

"""Extract some stats for OSM nodes, from a history OSM data file
"""

###############################################
# Import packages #############################
###############################################
import sys
import os.path as osp
from datetime import datetime
if sys.version_info[0] == 3:
    from datetime import timezone

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

import utils

###############################################
# Main method #################################
###############################################
if __name__ == '__main__':
    # call the script following
    # format 'python3 <data set name> <save_output?>' (2 args)
    if len(sys.argv) != 3:
        print("Usage: python3 <dataset_name> <save_output: y/n>")
        sys.exit(-1)
    dataset_name = sys.argv[1]
    save_output = True if sys.argv[2]=="y" or sys.argv[2]=="Y" else False
    print("Analyse of the OSM data gathered for {0}".format(dataset_name))
    if(save_output):
        print("Outputs will be saved into .csv files during process!")
    # Data reading
    tlnodes,tlways,tlrelations = utils.readOSMdata(dataset_name)
    print("There are {0} feature(s) and {1} individual(s) in node dataframe"
          .format(len(tlnodes.columns),len(tlnodes)))
    print("There are {0} feature(s) and {1} individual(s) in way dataframe"
          .format(len(tlways.columns),len(tlways)))
    print("There are {0} feature(s) and {1} individual(s) in relation dataframe"
          .format(len(tlrelations.columns),len(tlrelations)))

    ###############################################
    # Build last OSM elements starting from history data
    # and save them into dedicated files
    nodes = utils.updatedelem(tlnodes)
    ways = utils.updatedelem(tlways)
    relations = utils.updatedelem(tlrelations)
    if(save_output):
        utils.writeOSMdata(nodes,ways,relations,dataset_name)
    
    # Gather every elements into a single data frame
    elem = pd.Series(np.repeat(["n", "w", "r"]
                               , [len(tlnodes), len(tlways), len(tlrelations)]
                               , axis=0)
                     , name="elem").astype("category")
    osm_elements = pd.concat([tlnodes.loc[:,['id','version','visible','ts','uid'
                                             ,'chgset','ntags','tagkeys']],
                              tlways.loc[:,['id','version','visible','ts','uid'
                                            ,'chgset','ntags','tagkeys']],
                              tlrelations.loc[:,['id','version','visible','ts'
                                                 ,'uid','chgset','ntags'
                                                 ,'tagkeys']]]
                             , ignore_index=True)
    osm_elements = pd.concat([elem, osm_elements], axis=1)
    osm_elements = osm_elements.sort_values(by=['ts'])

    ###############################################
    # Analyse of elements
    userbyelem = osm_elements.groupby(
        ['elem','id'])['uid'].nunique().reset_index()
    versionbyelem = osm_elements.groupby(
        ['elem','id'])['version'].count().reset_index()
    elembegin = osm_elements.groupby(['elem','id'])['ts'].min().reset_index()
    elembegin.columns = ['elem','id','created_at']
    elemend = osm_elements.groupby(['elem','id'])['ts'].max().reset_index()
    elemend.columns = ['elem','id','lastmodif_at']
    elemsynthesis = pd.merge(userbyelem, versionbyelem, on=['elem','id'])
    elemsynthesis = pd.merge(elemsynthesis, elembegin, on=['elem','id'])
    elemsynthesis = pd.merge(elemsynthesis, elemend, on=['elem','id'])
    elemsynthesis['available'] = np.repeat([True], len(elemsynthesis))
    mask = np.logical_and(elemsynthesis.elem=="n",
                          elemsynthesis.id.isin(nodes.loc[nodes.visible==False
                                                          ,"id"]))
    elemsynthesis.loc[mask,"available"] = False
    mask = np.logical_and(elemsynthesis.elem=="w",
                          elemsynthesis.id.isin(ways.loc[ways.visible==False
                                                         ,"id"]))
    elemsynthesis.loc[mask,"available"] = False
    mask = np.logical_and(elemsynthesis.elem=="r",
                          elemsynthesis.id.isin(relations.loc[
                              relations.visible==False,"id"]))
    elemsynthesis.loc[mask,"available"] = False
    elemsynthesis['lifecycle'] = (elemsynthesis.lastmodif_at
                                  - elemsynthesis.created_at)
    timehorizon = pd.to_datetime("2017-02-13 00:00:00")
    elemsynthesis.loc[elemsynthesis.available==True,"lifecycle"] = timehorizon - elemsynthesis.loc[elemsynthesis.available==True,"created_at"]
    elemsynthesis.loc[elemsynthesis.available.values,"lifecycle"] = timehorizon - elemsynthesis.loc[elemsynthesis.available.values,"created_at"]

    ###############################################
    # Analyse of change sets
    chgsetsynthesis = (osm_elements.groupby(['chgset'])['elem']
                       .count().reset_index())
    chgsetsynthesis = pd.merge(chgsetsynthesis,
                               osm_elements.loc[:,['chgset','uid']]
                               .drop_duplicates(), on=['chgset'])
    nodemodifbychgset = (osm_elements.loc[osm_elements.elem=="n"]
                         .groupby(['chgset'])['elem'].count().reset_index())
    nodemodifbychgset.columns = ['chgset','nnode']
    chgsetsynthesis = pd.merge(chgsetsynthesis, nodemodifbychgset,
                               how="left", on=['chgset'])
    chgsetsynthesis.nnode[np.isnan(chgsetsynthesis.nnode)] = 0
    waymodifbychgset = (osm_elements.loc[osm_elements.elem=="w"]
                        .groupby(['chgset'])['elem'].count().reset_index())
    waymodifbychgset.columns = ['chgset','nway']
    chgsetsynthesis = pd.merge(chgsetsynthesis, waymodifbychgset,
                               how="left", on=['chgset'])
    chgsetsynthesis.nway[np.isnan(chgsetsynthesis.nway)] = 0
    relationmodifbychgset = (osm_elements.loc[osm_elements.elem=="r"]
                             .groupby(['chgset'])['elem'].count().reset_index())
    relationmodifbychgset.columns = ['chgset','nrelation']
    chgsetsynthesis = pd.merge(chgsetsynthesis, relationmodifbychgset,
                               how="left", on=['chgset'])
    chgsetsynthesis.nrelation[np.isnan(chgsetsynthesis.nrelation)] = 0
    chgsetbegin = osm_elements.groupby(['chgset'])['ts'].min().reset_index()
    chgsetbegin.columns = ['chgset','opened_at']
    chgsetend = osm_elements.groupby(['chgset'])['ts'].max().reset_index()
    chgsetend.columns = ['chgset','lastmodif_at']
    chgsetsynthesis = pd.merge(chgsetsynthesis, chgsetbegin, on=['chgset'])
    chgsetsynthesis = pd.merge(chgsetsynthesis, chgsetend, on=['chgset'])
    chgsetsynthesis['duration'] = (chgsetsynthesis.lastmodif_at
                                   - chgsetsynthesis.opened_at)
    
    ###############################################
    # Analyse of users
    usersynthesis = (osm_elements.groupby(['uid'])['chgset']
                     .nunique().reset_index())
    elembyuser = osm_elements.groupby(['uid'])['elem'].count().reset_index()
    usersynthesis = pd.merge(usersynthesis, elembyuser, on=['uid'])
    nodemodifbyuser = (osm_elements.loc[osm_elements.elem=="n"]
                       .groupby(['uid'])['elem'].count().reset_index())
    nodemodifbyuser.columns = ['uid','nnode']
    usersynthesis = pd.merge(usersynthesis, nodemodifbyuser,
                             how="left", on=['uid'])
    usersynthesis.nnode[np.isnan(usersynthesis.nnode)] = 0
    waymodifbyuser = (osm_elements.loc[osm_elements.elem=="w"]
                      .groupby(['uid'])['elem'].count().reset_index())
    waymodifbyuser.columns = ['uid','nway']
    usersynthesis = pd.merge(usersynthesis, waymodifbyuser,
                             how="left", on=['uid'])
    usersynthesis.nway[np.isnan(usersynthesis.nway)] = 0
    relationmodifbyuser = (osm_elements.loc[osm_elements.elem=="r"]
                           .groupby(['uid'])['elem'].count().reset_index())
    relationmodifbyuser.columns = ['uid','nrelation']
    usersynthesis = pd.merge(usersynthesis, relationmodifbyuser,
                             how="left", on=['uid'])
    usersynthesis.nrelation[np.isnan(usersynthesis.nrelation)] = 0
    contribbyuserelem = (osm_elements.groupby(['elem','id','uid'])
                         .size().reset_index())
    nmodifperuserperelem = (contribbyuserelem.groupby(['uid'])[0]
                            .median().reset_index())
    nmodifperuserperelem.columns = ['uid','nmodifperelem']
    usersynthesis = pd.merge(usersynthesis, nmodifperuserperelem,
                             how="left", on=['uid'])
    nnodemodifperuserperelem = contribbyuserelem.loc[contribbyuserelem.elem=="n"].groupby(['uid'])[0].median().reset_index()
    nnodemodifperuserperelem.columns = ['uid','nnodemodifperelem']
    usersynthesis = pd.merge(usersynthesis, nnodemodifperuserperelem,
                             how="left", on=['uid'])
    usersynthesis.nnodemodifperelem[np.isnan(usersynthesis.nnodemodifperelem)] = 0
    nwaymodifperuserperelem = contribbyuserelem.loc[contribbyuserelem.elem=="w"].groupby(['uid'])[0].median().reset_index()
    nwaymodifperuserperelem.columns = ['uid','nwaymodifperelem']
    usersynthesis = pd.merge(usersynthesis, nwaymodifperuserperelem,
                             how="left", on=['uid'])
    usersynthesis.nwaymodifperelem[np.isnan(usersynthesis.nwaymodifperelem)] = 0
    nrelationmodifperuserperelem = contribbyuserelem.loc[contribbyuserelem.elem=="r"].groupby(['uid'])[0].median().reset_index()
    nrelationmodifperuserperelem.columns = ['uid','nrelationmodifperelem']
    usersynthesis = pd.merge(usersynthesis, nrelationmodifperuserperelem,
                             how="left", on=['uid'])
    usersynthesis.nrelationmodifperelem[np.isnan(usersynthesis.nrelationmodifperelem)] = 0
    userbegin = osm_elements.groupby(['uid'])['ts'].min().reset_index()
    userbegin.columns = ['uid','first_at']
    userend = osm_elements.groupby(['uid'])['ts'].max().reset_index()
    userend.columns = ['uid','last_at']
    usersynthesis = pd.merge(usersynthesis, userbegin, on=['uid'])
    usersynthesis = pd.merge(usersynthesis, userend, on=['uid'])
    usersynthesis['activity'] = usersynthesis.last_at - usersynthesis.first_at

    ###############################################
    # Metadata saving
    if(save_output):
        utils.writeOSMmetadata(elemsynthesis, chgsetsynthesis,
                               usersynthesis, dataset_name)

    ###############################################
    # Plot metadata: number of unique contributors per element
    print("Plot 1: number of contributors per element...")
    y = np.linspace(0,1,1001)
    utils.single_empCDF(elemsynthesis.uid.quantile(y), y, lims=[0,20,0,1],
                        xlab="Number of unique contributors per OSM element",
                        ylab="Empirical CDF",
                        figpath="figs/uniquserperelem.png")
    print("Plot 2: number of contributors per element and per type...")
    x = elemsynthesis.groupby('elem')['uid'].quantile(y).unstack(0)
    utils.multi_empCDF(x, y, lims=[0,20,0,1], legend=True,
                       xlab="Number of unique contributors per OSM element",
                       ylab="Empirical CDF",
                       figpath="figs/uniquserperelemtype.png")

    # Plot metadata: number of versions per element
    print("Plot 3: number of versions per element...")
    utils.single_empCDF(elemsynthesis.version.quantile(y), y, lims=[0,20,0,1],
                        xlab="Number of versions per OSM element",
                        ylab="Empirical CDF",
                        figpath="figs/nbversionperelem.png")
    print("Plot 4: number of versions per element and per type...")
    x = elemsynthesis.groupby('elem')['version'].quantile(y).unstack(0)
    utils.multi_empCDF(x, y, lims=[0,80,0,1], legend=True,
                       xlab="Number of versions per OSM element",
                       ylab="Empirical CDF",
                       figpath="figs/nbversionperelemtype.png")

    # Plot metadata: element global description
    print("Plot 5: element global description...")
    utils.condsctrplot('uid', 'version', elemsynthesis.groupby('elem'),
                       lims=[0,150,0,600], bsctr=True, mkr="+", legend=True,
                       figpath="figs/elem_uniquservsversion.png",
                       xlab="Number of unique users per OSM element",
                       ylab="Number of version per OSM element")

    # Plot metadata: number of modifications per change set
    print("Plot 6: number of contributors per element...")
    utils.single_empCDF(chgsetsynthesis.elem.quantile(y), y, lims=[0,200,0,1],
                        hltqt=0.9, ylab="Empirical CDF",
                        xlab="Number of modifications per change set",
                        figpath="figs/nbmodifperchgset.png")
    print("Plot 7: number of contributors per element and per type...")
    x = chgsetsynthesis.loc[:,['nnode','nway','nrelation']].apply(
        lambda x: x.quantile(y))
    utils.multi_empCDF(x, y, lims=[0,100,0,1], hltqt=0.9, legend=True,
                       xlab="Number of modifications per change set",
                       ylab="Empirical CDF",
                       figpath="figs/nbmodifperchgsettype.png")
    
    # Plot metadata: change set global description
    print("Plot 8: change set global description...")
    x = chgsetsynthesis.loc[:,['nnode','nway','nrelation']]
    utils.multisctrplot(x, chgsetsynthesis.duration.astype('timedelta64[m]'),
                        lims=[0,np.log(max(x.apply(max))),0,700],
                        mkr="+", legend=True,
                        figpath="figs/chgset_modifvsduration.png",
                        xlab="Logarithm of modification quantity per change sets",
                        ylab="Change set duration (in minutes)")

    # Plot metadata: number of change sets per user
    print("Plot 9: number of change set per user...")
    utils.single_empCDF(usersynthesis.chgset.quantile(y), y, lims=[0,100,0,1],
                        xlab="Number of change sets per user",
                        ylab="Empirical CDF",
                        figpath="figs/nbchgsetperuser.png")
    
    # Plot metadata: number of modification per user
    print("Plot 10: number of modifications per user...")
    utils.single_empCDF(usersynthesis.elem.quantile(y), y,
                        lims=[0,100,0,1], hltqt=0.8,
                        xlab="Number of modifications per user",
                        ylab="Empirical CDF",
                        figpath="figs/nbmodifperuser.png")
    print("Plot 11: number of modifications per user and per type...")
    x = usersynthesis.loc[:,['nnode','nway','nrelation']].apply(
        lambda x: x.quantile(y))
    utils.multi_empCDF(x, y, lims=[0,200,0,1], hltqt=0.8, legend=True,
                       xlab="Number of modifications per user",
                       ylab="Empirical CDF",
                       figpath="figs/nbmodifperusertype.png")

    # Plot metadata: user global description
    print("Plot 12: user global description...")
    utils.scatterplot(usersynthesis.chgset,usersynthesis.elem, 
                      lims=[0,1000,0,10000], bsctr=True, mkr="s",
                      figpath="figs/user_chgsetvsmodif.png",
                      xlab="Number of change sets by user",
                      ylab="Number of modifications per user")
    print("Plot 13: number of contributors per element...")
    utils.scatterplot(usersynthesis.chgset,
                      usersynthesis.activity.astype('timedelta64[D]'),
                      lims=[0,100,0,3500], bsctr=True, mkr="+", 
                      figpath="figs/user_chgsetvsactivity.png",
                      xlab="Number of change sets by user",
                      ylab="User activity duration (in days)")
